# EDSL n Parameter Issue: Comprehensive Testing

This notebook demonstrates:
1. EDSL accepts but doesn't implement the `n` parameter
2. Direct API tests showing OpenAI (n≤128) and Gemini (candidateCount≤8) support
3. Cost implications: 99% savings on input tokens possible

**Update**: Added comprehensive testing of OpenAI and Gemini limits with exact boundaries.

In [1]:
# Setup
import os
import time
from edsl import Model, QuestionFreeText
from edsl.agents import Agent, AgentList

## Part 1: EDSL's n Parameter Doesn't Work

In [2]:
# Test EDSL with n=5
question = QuestionFreeText(
    question_name="number",
    question_text="Pick a random number between 1 and 100. Reply with just the number:"
)

model_with_n = Model('gpt-4o-mini', service_name='openai', n=5)

print("Testing EDSL with n=5 parameter...")
result = question.by(model_with_n).run()

results_list = result.to_list()
print(f"\nResults received: {len(results_list)}")
print(f"Expected: 5")
print(f"Status: {'✅ PASS' if len(results_list) == 5 else '❌ FAIL - n parameter not implemented'}")

Testing EDSL with n=5 parameter...

Results received: 1
Expected: 5
Status: ❌ FAIL - n parameter not implemented


## Part 2: OpenAI API - n Parameter Works (Limit: 128)

In [3]:
from openai import OpenAI

client = OpenAI(api_key=os.environ.get('OPENAI_API_KEY'))

print("Testing OpenAI n parameter directly:")
print("=" * 50)

# Test boundary values
test_values = [1, 10, 128, 129]

for n_val in test_values:
    print(f"\nn={n_val}:")
    try:
        response = client.chat.completions.create(
            model="gpt-4o-mini",
            messages=[{"role": "user", "content": "random number 1-100 nothing else"}],
            n=n_val,
            temperature=1.0,
            max_tokens=5
        )
        
        print(f"  ✅ SUCCESS: Got {len(response.choices)} completions")
        
        # Show examples for small n
        if n_val <= 10:
            numbers = [c.message.content.strip() for c in response.choices]
            print(f"  Numbers: {numbers}")
        
        # Show token usage
        print(f"  Input tokens: {response.usage.prompt_tokens} (charged once!)")
        print(f"  Savings: {(1 - 1/n_val)*100:.1f}% on input tokens")
        
    except Exception as e:
        if "maximum" in str(e).lower() or "Expected a value <= 128" in str(e):
            print(f"  ❌ LIMIT EXCEEDED: n must be ≤ 128")
        else:
            print(f"  ❌ Error: {str(e)[:100]}")

print("\n" + "=" * 50)
print("OpenAI n parameter limit: EXACTLY 128")

Testing OpenAI n parameter directly:

n=1:
  ✅ SUCCESS: Got 1 completions
  Numbers: ['42']
  Input tokens: 9 (charged once!)
  Savings: 0.0% on input tokens

n=10:
  ✅ SUCCESS: Got 10 completions
  Numbers: ['73', '42', '17', '88', '52', '91', '27', '64', '35', '79']
  Input tokens: 9 (charged once!)
  Savings: 90.0% on input tokens

n=128:
  ✅ SUCCESS: Got 128 completions
  Input tokens: 9 (charged once!)
  Savings: 99.2% on input tokens

n=129:
  ❌ LIMIT EXCEEDED: n must be ≤ 128

OpenAI n parameter limit: EXACTLY 128


## Part 3: Google Gemini - candidateCount Works (Limit: 8)

In [4]:
import google.generativeai as genai

# Configure Gemini
genai.configure(api_key='AIzaSyBimA7Xh0lRJKblUYY4dhStD31hVCV3TrQ')

print("Testing Gemini candidateCount parameter:")
print("=" * 50)

# Test with Gemini 2.5 Flash
model = genai.GenerativeModel('gemini-2.5-flash')

# Test boundary values
test_values = [1, 4, 8, 9]

for count in test_values:
    print(f"\ncandidateCount={count}:")
    try:
        config = genai.GenerationConfig(
            candidate_count=count,
            temperature=1.0,
            max_output_tokens=5
        )
        
        response = model.generate_content(
            "random number 1-100 nothing else",
            generation_config=config
        )
        
        print(f"  ✅ SUCCESS: Got {len(response.candidates)} candidates")
        
        # Show numbers
        numbers = []
        for c in response.candidates:
            if c.content and c.content.parts:
                numbers.append(c.content.parts[0].text.strip())
        if numbers:
            print(f"  Numbers: {numbers}")
        
        # Token usage
        if hasattr(response, 'usage_metadata'):
            print(f"  Input tokens: {response.usage_metadata.prompt_token_count} (charged once!)")
            print(f"  Savings: {(1 - 1/count)*100:.1f}% on input tokens")
        
    except Exception as e:
        if "[1, 8]" in str(e) or "must be in range" in str(e):
            print(f"  ❌ LIMIT EXCEEDED: candidateCount must be in [1, 8]")
        else:
            print(f"  ❌ Error: {str(e)[:100]}")

print("\n" + "=" * 50)
print("Gemini candidateCount limit: EXACTLY 8")

Testing Gemini candidateCount parameter:

candidateCount=1:
  ✅ SUCCESS: Got 1 candidates
  Numbers: ['72']
  Input tokens: 8 (charged once!)
  Savings: 0.0% on input tokens

candidateCount=4:
  ✅ SUCCESS: Got 4 candidates
  Numbers: ['37', '83', '19', '62']
  Input tokens: 8 (charged once!)
  Savings: 75.0% on input tokens

candidateCount=8:
  ✅ SUCCESS: Got 8 candidates
  Numbers: ['42', '17', '89', '53', '71', '26', '94', '38']
  Input tokens: 8 (charged once!)
  Savings: 87.5% on input tokens

candidateCount=9:
  ❌ LIMIT EXCEEDED: candidateCount must be in [1, 8]

Gemini candidateCount limit: EXACTLY 8


## Part 4: Provider Comparison Table

In [5]:
import pandas as pd

# Create comparison table
providers_data = [
    {"Provider": "OpenAI", "Parameter": "n", "Max Limit": 128, "Input Tokens": "Charged once", "Status": "✅ Tested"},
    {"Provider": "Azure OpenAI", "Parameter": "n", "Max Limit": 128, "Input Tokens": "Charged once", "Status": "Same as OpenAI"},
    {"Provider": "Google Gemini", "Parameter": "candidateCount", "Max Limit": 8, "Input Tokens": "Charged once", "Status": "✅ Tested"},
    {"Provider": "Anthropic", "Parameter": "None", "Max Limit": 0, "Input Tokens": "N/A", "Status": "❌ No support"},
    {"Provider": "Together AI", "Parameter": "None", "Max Limit": 0, "Input Tokens": "N/A", "Status": "❌ No support"},
    {"Provider": "Groq", "Parameter": "n (limited)", "Max Limit": 1, "Input Tokens": "N/A", "Status": "❌ n=1 only"},
    {"Provider": "Mistral", "Parameter": "None", "Max Limit": 0, "Input Tokens": "N/A", "Status": "❌ No support"},
]

df = pd.DataFrame(providers_data)
print("Provider Support for Multiple Completions:")
print("=" * 70)
print(df.to_string(index=False))

print("\n" + "=" * 70)
print("Key Findings:")
print("• OpenAI: Supports up to n=128 (16x more than Gemini)")
print("• Gemini: Supports up to candidateCount=8")
print("• Others: No support - require multiple API calls")
print("• Cost Savings: Up to 99% on input tokens with n=100")

Provider Support for Multiple Completions:
     Provider       Parameter Max Limit   Input Tokens        Status
       OpenAI               n       128   Charged once     ✅ Tested
 Azure OpenAI               n       128   Charged once Same as OpenAI
Google Gemini candidateCount         8   Charged once     ✅ Tested
    Anthropic            None         0            N/A  ❌ No support
  Together AI            None         0            N/A  ❌ No support
         Groq     n (limited)         1            N/A   ❌ n=1 only
      Mistral            None         0            N/A  ❌ No support

Key Findings:
• OpenAI: Supports up to n=128 (16x more than Gemini)
• Gemini: Supports up to candidateCount=8
• Others: No support - require multiple API calls
• Cost Savings: Up to 99% on input tokens with n=100


## Part 5: Cost Analysis - Real Impact

In [6]:
# Research scenario
num_prompts = 100
samples_per_prompt = 100
tokens_per_prompt = 150
tokens_per_completion = 50

# Pricing (GPT-4o-mini)
price_per_1k_input = 0.00015
price_per_1k_output = 0.00060

print("Cost Analysis: 100 prompts × 100 samples each")
print("=" * 60)

# Current EDSL approach
total_calls_current = num_prompts * samples_per_prompt
input_tokens_current = total_calls_current * tokens_per_prompt
output_tokens = total_calls_current * tokens_per_completion
cost_current = (input_tokens_current * price_per_1k_input + 
                output_tokens * price_per_1k_output) / 1000

print("\n📊 Current EDSL (no n parameter):")
print(f"  API calls: {total_calls_current:,}")
print(f"  Input tokens: {input_tokens_current:,}")
print(f"  Total cost: ${cost_current:.2f}")

# With OpenAI n parameter
total_calls_openai = num_prompts  # Just 1 call per prompt!
input_tokens_openai = total_calls_openai * tokens_per_prompt
cost_openai = (input_tokens_openai * price_per_1k_input + 
               output_tokens * price_per_1k_output) / 1000

print("\n✨ With OpenAI n=100:")
print(f"  API calls: {total_calls_openai:,}")
print(f"  Input tokens: {input_tokens_openai:,} (99% reduction!)")
print(f"  Total cost: ${cost_openai:.2f}")

# With Gemini candidateCount
total_calls_gemini = num_prompts * (samples_per_prompt // 8 + 1)
input_tokens_gemini = total_calls_gemini * tokens_per_prompt
cost_gemini = (input_tokens_gemini * price_per_1k_input + 
               output_tokens * price_per_1k_output) / 1000

print("\n🔷 With Gemini candidateCount=8:")
print(f"  API calls: {total_calls_gemini:,} (batched in 8s)")
print(f"  Input tokens: {input_tokens_gemini:,} (87% reduction)")
print(f"  Total cost: ${cost_gemini:.2f}")

print("\n" + "=" * 60)
print("💰 SAVINGS:")
print(f"  OpenAI vs Current: ${cost_current - cost_openai:.2f} saved ({(1-cost_openai/cost_current)*100:.0f}% cheaper)")
print(f"  Gemini vs Current: ${cost_current - cost_gemini:.2f} saved ({(1-cost_gemini/cost_current)*100:.0f}% cheaper)")

Cost Analysis: 100 prompts × 100 samples each

📊 Current EDSL (no n parameter):
  API calls: 10,000
  Input tokens: 1,500,000
  Total cost: $0.52

✨ With OpenAI n=100:
  API calls: 100
  Input tokens: 15,000 (99% reduction!)
  Total cost: $0.30

🔷 With Gemini candidateCount=8:
  API calls: 1,300 (batched in 8s)
  Input tokens: 195,000 (87% reduction)
  Total cost: $0.33

💰 SAVINGS:
  OpenAI vs Current: $0.23 saved (43% cheaper)
  Gemini vs Current: $0.20 saved (37% cheaper)


## Summary

### Confirmed Limits (Empirically Tested)
- **OpenAI**: n ≤ 128 (n=129 fails with "Expected a value <= 128")
- **Gemini**: candidateCount ≤ 8 (9 fails with "must be in range [1, 8]")

### Cost Impact
- For 100 samples: **99% savings** with OpenAI, **87.5% savings** with Gemini
- Research with 10,000 total completions: Save **$0.23** with proper implementation

### EDSL Issues Filed
- Issue #2185: n parameter not implemented
- Issue #2184: No parameter validation (accepts invalid params)
- PR #2186: Added parameter validation system

### Next Steps
Implement proper n parameter support in EDSL's `run()` method to leverage these native API features.